File: lee/src/main/java/org/example/lee/题目/矩阵/b_螺旋矩阵.java
package org.example.lee.题目.矩阵;

import java.util.ArrayList;
import java.util.List;


public class b_螺旋矩阵 {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length,n = matrix[0].length;
        List<Integer> list = new ArrayList<>();
        int a=0,i=0,j=0;
        while(list.size()<m*n){
            for (i=a,j = a; j < n-a; j++) {
                list.add(matrix[i][j]);
            }
            for (i++,j--; i < m-a;i++){
                list.add(matrix[i][j]);
            }
            if(list.size()>=m*n){
                return list;
            }
            for (i--,j-- ; j > a; j--) {
                list.add(matrix[i][j]);
            }
            for (; i >a;i--){
                list.add(matrix[i][j]);
            }
            a++;
        }
        return list;
    }
}

File: lee/src/main/java/org/example/lee/题目/矩阵/b_螺旋矩阵.go
package 矩阵

// 自己模拟下规律就好
func spiralOrder(matrix [][]int) []int {
	var arr []int
	var a int
	lenI, lenJ := len(matrix), len(matrix[0])
	q, p, w := 0, lenI, lenJ
	for a < len(matrix)*len(matrix[0]) {
		for i := q; i < w; i++ {
			arr = append(arr, matrix[q][i])
			a++
		}
		for i := q + 1; i < p; i++ {
			arr = append(arr, matrix[i][w-1])
			a++
		}
		if a >= len(matrix)*len(matrix[0]) {
			break
		}
		for i := w - 2; i >= q; i-- {
			arr = append(arr, matrix[p-1][i])
			a++
		}
		for i := p - 2; i >= q+1; i-- {
			arr = append(arr, matrix[i][q])
			a++
		}
		q++
		p--
		w--
	}
	return arr
}
